text/0000-externally-implementable-traits.md

```diff
@@ -0,0 +1,251 @@
+
+## Signatures
+
+As with a normal trait `impl`, the signatures of functions in the `extern impl` must match those defined in the `extern trait`.
+
+(As with normal traits, whether `#[track_caller]` is used or not is considered part of the signature here.)
+
+## Implementations
+
+An `extern impl` must provide implementations for all functions defined in an `extern trait`, except those for which a default implementation is provided.
+
+It is an error to have no `extern impl` item (in any crate) for an `extern trait` item, unless *all* functions in the `extern trait` have a default implementation.
+
+It is an error to have multiple `extern impl` items (across all crates) for the same `extern trait` item.
+
+Note: This means that adding or removing an `extern impl` item is a semver incompatible change.
+
+## Visibility
+
+`extern trait` items can have a visibility specifier (like `pub`), which determines who can *call* functions in the trait (or create pointers to it, etc.). Individual functions in the trait may not have a visibility modifer.
+
+*Implementing* the trait can be done by any crate that can name the item.
+(The `extern impl` item will need to name the item to implement, which could be directly or through an alias/re-export.)
+
+# Implementation
+
+The implementation will be based on the same mechanisms as used today for the `panic_handler` and `#[global_allocator]` features.
+
+The compiler of the root crate will find the implementation of all externally implementable traits and give an error
+if more than one implementation is found for any of them.
+If none are found, the result is either an error, or—if the all functions in an `extern trait` have a default body—an implementation is generated that calls those default bodies.
+
+# Drawbacks
+
+- It encourages globally defined behaviour.
+  - Counterargument: We are already doing this anyway, both inside the standard library (e.g. panic_handler, allocator)
+    and outside (e.g. global logger). This just makes it much easier (and safer) to get right.
+- This will invite the addition of many hooks to the standard library to modify existing behavior.
+  While we should consider such possibilities, this RFC does not propose that every piece of standard library behavior should be replaceable.
+
+# Rationale and alternatives
+
+## Syntax
+
+The syntax re-uses existing keywords. Alternatively, we could:
+  - Use the `override` reserved keyword.
+  - Add a new (contextual) keyword (e.g. `extern existential`).
+  - Use an attribute (e.g. `#[extern_impl]`) instead.
+
+## Functions or statics
+
+This RFC only proposes externally implementable *functions*.
+
+An alternative is to only provide externally definable *statics* instead.
+
+That would be equivalent in power: one can store a function pointer in a static, and one can return a reference to a static from a function ([RFC 3635](https://github.com/rust-lang/rfcs/pull/3635)).
```

Comment:
They are equivalent in power, but not necessarily in ergonomics — not how they invite themselves to be used. I think there's a serious problem lurking here.

One of the benefits of the way [`#[global_allocator]`](https://doc.rust-lang.org/reference/runtime.html#the-global_allocator-attribute) works is that a library can trivially provide a type which implements `GlobalAlloc` (let's call it `foo::Alloc`), but leave it to a dependent crate (let's call it `bar`) to actually declare the static item that uses that type. That dependent can therefore choose to, for example, write its own type `bar::Alloc` which **wraps** `foo::Alloc` to modify its behavior.

On the other hand, the global allocation interface in Example 2 in this RFC does not have this virtue: the easiest path (in the absence of macro support) is for `foo` to write `unsafe extern impl GlobalAllocator { /* several function definitions */ }`, and then `bar` cannot extend `foo`'s allocator implementation. **`foo` would have to go out of its way** to provide an allocator that can be wrapped.

`foo` can already make itself unusable to `bar` by writing a `#[global_allocator]` attribute, but the advantage of the declaration being a single static instead of a whole `impl` block is that it's *easy* for `foo` to not do that, and instead just tell its users "declare this static to activate my implementation"; it's *less code* than the less flexible option. I worry that if this RFC were implemented as-is,

* crates defining externally implementable traits would write simple-looking traits (rather than ones with extra indirection through another type like Example 3), and
* crates implementing them would do so in the least-effort way: just write the `extern impl` with code inside it, providing no opportunity for dependents to build on that implementation.

It's *possible* for the trait defining crate to mitigate this — such as by providing the macro suggested in Example 2, or writing a specifically single-function trait like Example 3. But that's a mitigation by writing more code. I think this feature needs to be designed so that writing straightforward minimal code has good results — and externally definable statics (that worked like `#[global_allocator]` currently does) would do that.

(None of this issue is unique to allocators; it applies to any interface made up of multiple functions, such as `log::Log`. I'm focusing on the allocator example because it has multiple methods, it's already in the RFC, and we already have `#[global_allocator]` to compare to. I've wanted or required this type of flexibility for both allocators (in test instrumentation) and loggers (to send logs to multiple destinations).)